File: c/triangle/triangle.c
#include "triangle.h"

static int is_valid_triangle(triangle_t sides) {
    if (sides.a <= 0 || sides.b <= 0 || sides.c <= 0)
        return 0;
    return sides.a + sides.b >= sides.c && sides.b + sides.c >= sides.a && sides.c + sides.a >= sides.b;
}

static int unique_sides(triangle_t sides) {
    if (sides.a == sides.b) return (sides.c == sides.a) ? 1 : 2;
    return (sides.b == sides.c || sides.a == sides.c) ? 2 : 3;
}

int is_equilateral(triangle_t sides) {
    return is_valid_triangle(sides) && unique_sides(sides) == 1;
}

int is_isosceles(triangle_t sides) {
    return is_valid_triangle(sides) && unique_sides(sides) < 3;
}

int is_scalene(triangle_t sides) {
    return is_valid_triangle(sides) && unique_sides(sides) == 3;
}

File: zig/triangle/triangle.zig
const std = @import("std");

pub const TriangleError = error{Invalid};

pub const Triangle = struct {
    sides: [3]f64,
    pub fn init(a: f64, b: f64, c: f64) TriangleError!Triangle {
        var sides = [3]f64{ a, b, c };
        std.sort.sort(f64, &sides, {}, comptime std.sort.asc(f64));
        if (sides[0] + sides[1] <= sides[2] or sides[0] <= 0) {
            return TriangleError.Invalid;
        }
        return Triangle{ .sides = sides };
    }
    pub fn isEquilateral(self: Triangle) bool {
        return self.sides[0] == self.sides[2];
    }
    pub fn isIsosceles(self: Triangle) bool {
        return self.sides[0] == self.sides[1] or self.sides[1] == self.sides[2];
    }
    pub fn isScalene(self: Triangle) bool {
        return self.sides[0] != self.sides[1] and self.sides[1] != self.sides[2];
    }
};
